src/rpcserver.cpp

```diff
@@ -0,0 +1,264 @@
+                    }
+                }
+                else
+                {
+                    Send ( pSocket, QJsonDocument ( CreateJsonRpcErrorReply ( iErrInvalidRequest, "Invalid request: null is not allowed" ) ) );
+                }
+            }
+            else if ( data.isArray() )
+            {
+                // JSON-RPC batch mode: multiple requests in an array
+                QJsonArray output;
+                for ( auto item : data.array() )
+                {
+                    if ( !item.isObject() )
+                    {
+                        output.append (
+                            CreateJsonRpcErrorReply ( iErrInvalidRequest, "Invalid request: Non-object item encountered in a batch request array" ) );
+                        continue;
+                    }
+                    auto        object = item.toObject();
+                    QJsonObject response;
+                    response["jsonrpc"] = QJsonValue ( "2.0" );
+                    response["id"]      = object["id"];
+                    ProcessMessage ( pSocket, object, response );
+                    output.append ( response );
+                }
+                if ( output.size() > 0 )
+                {
+                    Send ( pSocket, QJsonDocument ( output ) );
+                }
+                else
+                {
+                    Send ( pSocket,
+                           QJsonDocument ( CreateJsonRpcErrorReply ( iErrInvalidRequest, "Invalid request: Empty batch request encountered" ) ) );
+                }
+            }
+            else if ( data.isObject() )
+            {
+                auto        object = data.object();
+                QJsonObject response;
+                response["jsonrpc"] = QJsonValue ( "2.0" );
+                response["id"]      = object["id"];
+                ProcessMessage ( pSocket, object, response );
+                Send ( pSocket, QJsonDocument ( response ) );
+            }
+            else
+            {
+                Send ( pSocket,
+                       QJsonDocument (
+                           CreateJsonRpcErrorReply ( iErrInvalidRequest,
+                                                     "Invalid request: Unrecognized JSON; a request must be either an object or an array" ) ) );
+            }
+        }
+    } );
+}
+
+void CRpcServer::Send ( QTcpSocket* pSocket, const QJsonDocument& aMessage ) { pSocket->write ( aMessage.toJson ( QJsonDocument::Compact ) + "\n" ); }
+
+void CRpcServer::HandleApiAuth ( QTcpSocket* pSocket, const QJsonObject& params, QJsonObject& response )
+{
+    auto userSecret = params["secret"];
+    if ( !userSecret.isString() )
+    {
+        response["error"] = CreateJsonRpcError ( iErrInvalidParams, "Invalid params: secret is not a string" );
+        return;
+    }
+
+    if ( userSecret == strSecret )
+    {
+        isAuthenticated[pSocket] = true;
+        response["result"]       = "ok";
+        return;
```

Comment:
```suggestion
        qInfo() << "- JSON-RPC: accepted valid authentication secret from" << pSocket->peerAddress().toString();
        return;
```

To log successful logins too (might make security hardening easier)

Reply:
> To log successful logins too (might make security hardening easier)

On every request?  That's going to be rather noisy in the log.  It's definitely not needed as an info-level message, I'd say.

Reply:
Is this done on every request?

I don't think so. It's just done on every api-auth call, I think.

qDebug - in my opinion is not "strong" enough. It's certainly a "note" not an error but also not something "unimportant".

Reply:
It's once per connection (there could be multiple apiAuth calls per connection, but that would not be necessary nor recommended). Each connect already generates a log message anyway.
The number of connection attempts depends on the API client. There might be long-lived connections where the number of (re)connects, and as such the number of log messages, would be low. Other usages might be stateless and would rather connect on demand (thinking of PHP-based backends, for example). That would generate more log noise.

I think it might be OK to add this...?

Reply:
@hoffie I'd still like this to be included. Waiting on opinions of others. Hopefully it's ready to merge then.

Reply:
Yep.

> Hopefully it's ready to merge then.

The whole API docs stuff is still missing (see dtinth's latest comment), I fear.

Reply:
> The whole API docs stuff is still missing (see dtinth's latest comment), I fear.

But that's easily creatable?

Reply:
Sure, I didn't want to say that it's a blocker, just that there's still other TODOs left.

Reply:
Yeah. I'd really like to have this merged someday soon since it gets really difficult to follow (and is in a working state).

Reply:
Anyone against adding this suggestion?

Reply:
For what its worth, I agree it is getting very difficult to follow this PR.  I think marking as experimental even if there are open ‘todo’s’ and getting it into master so others can really start to use it in context makes the most sense.